[WIP][analyzer] Refactor `ExplodedGraph::trim()`

NagyDonat opened the pull request:
...because its implementation was using a complicated two-pass graph traversal while in fact one pass is sufficient for performing that trimming operation.

Note that this commit changes the interface of `trim`: it no longer accepts nullpointers within the array `Sinks` and `BugReporter.cpp` (the single non-debug caller) was modified accordingly. This also affects the interface of `DumpGraph` (which calls `trim` under the hood), but that's a debug helper which is only called if some developer compiles a tweaked version of the analyzer for local debugging and it is unlikely that a developer would pass nullpointers to it.

WORK IN PROGRESS -- DO NOT MERGE!

After this commit the analyzer will pick different representatives from the bug report equivalence classes (because some iteration order changes), which breaks lots of testcases.

NagyDonat committed:
[WIP][analyzer] Refactor `ExplodedGraph::trim()`

...because its implementation was using a complicated two-pass graph
traversal while in fact one pass is sufficient for performing that
trimming operation.

Note that this commit changes the interface of `trim`: it no longer
accepts nullpointers within the array `Sinks` and `BugReporter.cpp` (the
single non-debug caller) was modified accordingly. This also affects the
interface of `DumpGraph` (which calls `trim` under the hood), but that's
a debug helper which is only called if some developer compiles a tweaked
version of the analyzer for local debugging and it is unlikely that a
developer would pass nullpointers to it.

WORK IN PROGRESS -- DO NOT MERGE!

After this commit the analyzer will pick different representants from
the bug report equivalence classes (because some iteration order
changes), which breaks lots of testcases.

NagyDonat commented:
This is the cleanup of `ExplodedGraph::trim()` that I promised at https://github.com/llvm/llvm-project/commit/31e981ca1dc323c8a32012cb60a0a8fe3985db1a. This significantly simplifies the algorithm and should be equivalent in a theoretical sense, but unfortunately it changes the (annoyingly fragile) ordering of reports within a bug report equivalence class :disappointed:

@steakhal What should  I do with this?

NagyDonat committed:
Simplify map lookup; unbrace

NagyDonat commented on clang/lib/StaticAnalyzer/Core/ExplodedGraph.cpp:
```diff
@@ -442,109 +442,65 @@ std::unique_ptr<ExplodedGraph>
-        continue;
+      auto Iterator = ForwardMap->find(Succ);
+      if (Iterator != ForwardMap->end()) {
+        const_cast<ExplodedNode *>(Iterator->second)->addPredecessor(NewN, *Trimmed);
```
Thanks for the suggestion, I simplified these two blocks.

NagyDonat committed:
Get rid of InverseMap because it was unused

NagyDonat committed:
Eliminate const_cast by using the proper type

steakhal commented on clang/lib/StaticAnalyzer/Core/ExplodedGraph.cpp:
```diff
@@ -479,25 +479,20 @@ ExplodedGraph::trim(ArrayRef<const NodeTy *> Sinks,
-      if (Iterator != ForwardMap->end()) {
-        NewN->addPredecessor(const_cast<ExplodedNode *>(Iterator->second), *Trimmed);
-      } else {
+      if (const ExplodedNode *Mapped = ForwardMap->lookup(Pred))
+        NewN->addPredecessor(const_cast<ExplodedNode *>(Mapped), *Trimmed);
```
Given these const_cast, I wonder if we could just put mutable pointers to the `ForwardMap` so that we wouldn't need to cast it away from these places. WDYT? but its really bikeshedding at this point. I'm happy as we have this already.

NagyDonat committed:
Avoid copying a vector of node pointers

NagyDonat committed:
Delete an unused type alias

NagyDonat committed:
Capitalize the parameter 'bugReports'

NagyDonat committed:
Clarify that PathSensitiveBugReport::ErrorNode is never null

The declaration of this data member was specifying `nullptr` as an
initializer, but each constructor of the class initializes it to a
non-null value (which is taken as an argument) and nothing changes its
value later.

NagyDonat committed:
Fix misleading declaration of a loop variable

It looks like an iterator, let's clarify its type and use a better name.

NagyDonat committed:
Satisfy git-clang-format

NagyDonat commented on clang/lib/StaticAnalyzer/Core/BugReporter.cpp:
```diff
@@ -2633,7 +2633,8 @@ BugPathGetter::BugPathGetter(const ExplodedGraph *OriginalGraph,
            "We only allow BugReporterVisitors and BugReporter itself to "
            "invalidate reports!");
-    Nodes.emplace_back(I->getErrorNode());
+    if (const ExplodedNode *ErrNode = I->getErrorNode())
```
Yes, I looked into it and you're right, it cannot be null. I eliminated the conditional and clarified the situation in commit https://github.com/llvm/llvm-project/pull/139939/commits/f6b4ef520af22f8437b8b6423e68dcfae81c8690 . Thanks for the tip!

NagyDonat committed:
Satisfy git-clang-format (really)

Previous call only checked the last few commits from the PR.

steakhal commented on clang/lib/StaticAnalyzer/Core/ExprEngine.cpp:
```diff
@@ -4096,7 +4096,8 @@ std::string ExprEngine::DumpGraph(bool trim, StringRef Filename) {
 std::string ExprEngine::DumpGraph(ArrayRef<const ExplodedNode *> Nodes,
                                   StringRef Filename) {
-  std::unique_ptr<ExplodedGraph> TrimmedG(G.trim(Nodes));
+  TrimGraphWorklist Worklist{Nodes};
+  std::unique_ptr<ExplodedGraph> TrimmedG(G.trim(Worklist));
```
Why is `Worklist` hoised into a variable?

steakhal replied:
Ah because it's now an in-out ref parameter. I don't understand why it needs to be an lvalue.

NagyDonat replied:
Oh, I see... I can just pass `Nodes` directly to `G.trim` and the type conversion would force the materialization of an unnamed temporary object, which is exactly the right thing in this situation.

steakhal replied:
One should prefer views at parameter boundaries.
You can take an arrayref parameter type, and then inside your function materialize a small vector out of it to use it as a worklist.

NagyDonat commented:
I pushed a heap of small commits which do various simple cleanup in this neighborhood.

--------------------

> If we could harden the bug EQclass part for deterministically selecting the sequence of the bug reports we would try, that would make the outcomes stable no matter how the egraph looks, right?
>
> Anyway, we should probably focus first hardening stability of the reports and then on disruptive changes to mitigate the pain upfront. Because if we would land this, and then more similar changes, then we would face this instability again and again with a similar magnitude.

I agree, this seems to be a good approach: first land one commit that enforces deterministic ordering of the bug reports (which is non-NFC and may disrupt the users _once_) and then it will be possible to merge this commit and other similar changes without further disruptions.

I guess, then this commit is put on hold until you or me or somebody else can harden the ordering of bug reports within an EQclass... I'm not sure that I can do so immediately (my backlog of plans is getting larger each day...) but there is a chance...

steakhal commented:
> I pushed a heap of small commits which do various simple cleanup in this neighborhood.
> 
> > If we could harden the bug EQclass part for deterministically selecting the sequence of the bug reports we would try, that would make the outcomes stable no matter how the egraph looks, right?
> > Anyway, we should probably focus first hardening stability of the reports and then on disruptive changes to mitigate the pain upfront. Because if we would land this, and then more similar changes, then we would face this instability again and again with a similar magnitude.
> 
> I agree, this seems to be a good approach: first land one commit that enforces deterministic ordering of the bug reports (which is non-NFC and may disrupt the users _once_) and then it will be possible to merge this commit and other similar changes without further disruptions.
> 
> I guess, then this commit is put on hold until you or me or somebody else can harden the ordering of bug reports within an EQclass... I'm not sure that I can do so immediately (my backlog of plans is getting larger each day...) but there is a chance...

I think we are aligned. yes.

NagyDonat committed:
Provide 'ExplodedGraph::trim(ArrayRef, ...)'

NagyDonat commented:
I started a measurement on open source projects to see the effect of this change on the (total) analysis runtime. I don't expect much but if this turns out to be non-negligible, then I will prioritize this direction.

steakhal commented:
> I started a measurement on open source projects to see the effect of this change on the (total) analysis runtime. I don't expect much but if this turns out to be non-negligible, then I will prioritize this direction.

RT is not a concern to me in this case.

NagyDonat commented:
> RT is not a concern to me in this case.

The reason why I decided to check the runtime is because you said that "It not only simplifies it but also makes if faster. " in an earlier commit.

My first runtime measurement finished and unfortunately it seems that this PR somehow _increases_ the runtime :confused: on average:

| Project | before this PR| this PR | Ratio |
| --- | --- | --- | --- |
| memcached | 23.78 | 26.44 | 111.2% |
| tmux | 74.92 | 73.93 | 98.7% |
| curl | 231.98 | 238.36 | 102.8% |
| twin | 57.52 | 54.77 | 95.2% |
| vim | 179.32 | 164.70 | 91.8% |
| openssl | 407.68 | 407.90 | 100.1% |
| sqlite | 878.11 | 835.58 | 95.2% |
| ffmpeg | 1306.35 | 1288.38 | 98.6% |
| postgres | 401.11 | 397.88 | 99.2% |
| tinyxml2 | 28.74 | 28.44 | 98.9% |
| libwebm | 57.17 | 60.29 | 105.5% |
| xerces | 204.98 | 204.06 | 99.6% |
| bitcoin | 160.97 | 158.15 | 98.2% |
| protobuf | 2332.42 | 2435.38 | 104.4% |
| qtbase | 3226.83 | 3851.08 | 119.3% |
| openrct2 | 1154.17 | 1172.49 | 101.6% |
| **total** | 10726.05 | 11397.83 | 106.3% |

I restarted this measurement to see whether this is a real tendency or just noise... but if this is a real slowdown, then I'll probably abandon this direction.

steakhal commented:
I don't believe these numbers. The report processing itself should barely take a measurable time.
Improving that shouldn't be visible. I suspect what you measured was noise.
I wouldn't invest into further measurements.

steakhal commented:
Thanks for letting me know.
